### DISCUSSION TITLE
en/chapter_data_structure/classification_of_data_structure/

### DISCUSSION THREAD

**Discussion by giscus** [krahets/hello-algo#1529] [Announcements]
# en/chapter_data_structure/classification_of_data_structure/

Data Structures and Algorithms Crash Course with Animated Illustrations and Off-the-Shelf Code

https://www.hello-algo.com/en/chapter_data_structure/classification_of_data_structure/

**Comment by rtau** [comment]
Hash table is listed under both linear and nonlinear data structure, what does it mean?

**Reply by Kyrie666** [reply]
Hash tables are a data structure that allows for fast insertion, deletion, and search operations. They typically use a hash function to calculate the storage location for data elements. Due to this rapid access characteristic, which usually does not depend on the order of the data elements, hash tables are considered nonlinear data structures.

However, hash tables can also be seen as linear data structures because, at the underlying level, they often use arrays to store data elements, and arrays are a linear data structure. In an array, data elements are stored in a linear sequence.

So, when it is said that hash tables are categorized under both linear and nonlinear data structures, it means that they combine the characteristics of these two types of data structures: logically, they provide rapid access capabilities, behaving as nonlinear structures; while physically, they might use linear structures like arrays to organize data. This is why hash tables can be classified as both linear and nonlinear data structures.

**Reply by krahets** [reply]
In terms of “logical structure,” we typically categorize hash tables as a linear data structure. However, regarding their “physical structure,” hash tables can incorporate both linear (arrays) and non-linear (often tree-based) data structures. How we categorize depends on the perspective we choose to observe from.

**Comment by jessicakfl** [comment]
Is it any relationship between data structure and algorithm type when problem solving?
